watertap/unit_models/crystallizer.py

```diff
@@ -0,0 +1,696 @@
+
+        @self.Constraint()    
+        def eq_p_con2(b):
+            return b.properties_in[0].pressure - b.properties_solids[0].pressure == 0
+
+        @self.Constraint()    
+        def eq_p_con3(b):
+            return b.properties_vapor[0].pressure - self.pressure_operating == 0
+
+        @self.Constraint()    
+        def eq_T_con1(b):
+            return self.temperature_operating - b.properties_solids[0].temperature == 0
+
+        @self.Constraint()    
+        def eq_T_con2(b):
+            return self.temperature_operating - b.properties_vapor[0].temperature == 0
+
+        @self.Constraint()
+        def eq_T_con3(b):
+            return self.temperature_operating - b.properties_out[0].temperature == 0
+
+
+        # 7. Heat exchanger minimum circulation flow rate calculations - see Lewis et al. or Tavare et al.
+        @self.Constraint(doc="Constraint on mimimum circulation rate through crystallizer heat exchanger")
+        def eq_minimum_hex_circulation_rate_constraint(b):
+            rho_cp_dt = self.hex_max_temp_increase * \
+            (
+                b.product_volumetric_solids_fraction * b.properties_solids[0].dens_mass_solute['Sol'] * b.properties_solids[0].cp_solute['Sol'] 
+                + (1 - b.product_volumetric_solids_fraction) * b.properties_out[0].dens_mass_phase['Liq'] * b.properties_out[0].cp_phase['Liq']
+                )
+            return (
+                b.magma_circulation_flow_vol * rho_cp_dt == pyunits.convert(b.work_mechanical[0], to_units=pyunits.J / pyunits.s)
+                )
+
+        # 8. Suspension density
+        @self.Constraint(doc="Slurry density calculation")
+        def eq_slurry_density(b):
+            return (self.slurry_density == 
+                b.product_volumetric_solids_fraction * b.properties_solids[0].dens_mass_solute['Sol'] 
+                + (1 - b.product_volumetric_solids_fraction) * b.properties_out[0].dens_mass_phase['Liq']
+                )
+
+
+        # 9. Residence time calculation
+        @self.Constraint(doc="Residence time")
+        def eq_residence_time(b):
+            return (
+                b.t_res == b.L_d / (3.67 * pyunits.convert(b.G_rate, to_units=pyunits.m/pyunits.hr))
+                )
+
+        # 10. Suspension volume calculation
+        @self.Constraint(doc="Suspension volume")
+        def eq_suspension_volume(b):
+            return (
+                b.V_suspension == (b.properties_solids[0].flow_vol + b.properties_out[0].flow_vol) * pyunits.convert(b.t_res, to_units=pyunits.s)
+                )
+
+
+        # 11. Minimum diameter of evaporation zone
+        def eq_max_allowable_velocity(b):
+            return b.k_param * \
+            (b.properties_out[0].dens_mass_phase['Liq']/ b.properties_vapor[0].dens_mass_solvent['Vap']) ** 0.5
+        
+        self.v_max = Expression(rule=eq_max_allowable_velocity, doc='maximum allowable vapour linear velocity in m/s') 
```

Comment:
You can use the same decorator syntax you've been using for constraints for an expression.

Reply:
Done.